File: sdk/src/main/java/com/cxense/cxensesdk/model/UserIdentity.java
package com.cxense.cxensesdk.model;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * User identity object.
 *
 * @author Dmitriy Konopelkin (dmitry.konopelkin@cxense.com) on (2017-06-05).
 */
@SuppressWarnings({"UnusedDeclaration", "WeakerAccess"}) // Public API.
public class UserIdentity extends BaseUserIdentity {
    protected UserIdentity() {
        super();
    }

    protected UserIdentity(@Nullable String type) {
        super(type);
    }

    /**
     * Creates new instance with user identifier value and user identifier type.
     *
     * @param id   user identifier
     * @param type user identifier type
     */
    public UserIdentity(@Nullable String id, @Nullable String type) {
        super(id, type);
    }

    /**
     * Creates new instance from another {@link UserIdentity}
     *
     * @param identity UserIdentity instance
     */
    public UserIdentity(@NonNull UserIdentity identity) {
        this(identity.id, identity.type);
    }
}

File: sdk/src/main/java/com/cxense/cxensesdk/model/UserIdentity.kt
package com.cxense.cxensesdk.model

import com.google.gson.annotations.SerializedName

/**
 * User identity object.
 * @param type user identifier type
 * @property type user identifier type
 * @property id user identifier
 */
open class UserIdentity constructor(
    type: String,
    @SerializedName("id") val id: String
) {
    @SerializedName("type")
    val type: String = type.also {
        require(it == CX_USER_TYPE || it.matches(USER_TYPE_REGEX.toRegex())) {
            """Type should be "cx" or a three character lower-case alpha-numeric string"""
        }
    }

    companion object {
        private const val CX_USER_TYPE = "cx"
        private const val USER_TYPE_REGEX = "\\w{3}"
    }
}
